docs/design/features/host-tracing.md

```diff
@@ -1,87 +1,96 @@
-In case of new (.NET Core 3) `hostfxr` component which would call into an old (.NET Core 2.1) `hostpolicy` component, the `hostfxr` will not perform the propagation in any way since the older `hostpolicy` doesn't support this mechanism.
+The `hostfxr_set_error_writer` sets the error writer only for the current thread (the setting is thread local). The custom host must set it separately (and potentially with different callbacks) on each thread it wants to use `hostfxr` functions (and have them redirect errors).
 
-The trace listener interface looks like this:
-``` C++
-struct host_trace_listener
-{
-    void trace_verbose(const pal::char_t *message, const pal::char_t *activityId);
-    void trace_info(const pal::char_t *message, const pal::char_t *activityId);
-    void trace_warning(const pal::char_t *message, const pal::char_t *activityId);
-    void trace_error(const pal::char_t *message, const pal::char_t *activityId);
-    void flush();
-}
+Only one error writer can be registered on a given thread at any given time.
+
+All errors are also written to a trace output if one is enabled (via `COREHOST_TRACE=1`) regardless of which error routing is used.
+
+The `hostfxr` component propagates the error writer to the `hostpolicy` component before it calls into it. So custom host only needs to register its error writer with the `hostfxr` component. The propagation of the error writer is only done for the duration necessary after which it will be unregistered again.
+In case of a .NET Core 3+ `hostfxr` which would call into an old (.NET Core 2.1) `hostpolicy` component, the `hostfxr` will not perform the propagation in any way since the older `hostpolicy` doesn't support this mechanism.
+
+The error writer callback is declared as:
+
+```cpp
+typedef void (__cdecl *error_writer_fn)(const pal::char_t* message);
 ```
 
-The `message` parameter is a standard `NUL` terminated string and it's the message to trace with the respective verbosity level.
-The `activityId` parameter is a standard `NUL` terminated string. It's used to correlate traces for a given binding event. The content of the string is not yet defined, but the trace listeners should consider it opaque. Trace listeners should include this string in the trace of the message in some form. The parameter may be `NULL` in which case the trace doesn't really belong to any specific binding event.
+The `message` parameter is a standard `NUL` terminated string. The memory for it is owned by the caller (so some of the hosting components, it may not be just `hostfxr`) and it's only valid for the duration of the call to the error writer callback.
```

Comment:
```suggestion
The `message` parameter is a standard `NULL` terminated string. The memory for it is owned by the caller (so some of the hosting components, it may not be just `hostfxr`) and it's only valid for the duration of the call to the error writer callback.
```